---
title: "MVMR"
output: html_document
date: "2023-01-26"
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '~/gitcode/MR-tutorial')
```

## Load R packages 
```{r packages}
library(tidyverse)    # Data wrangling 
library(cowplot)
library(TwoSampleMR)  # MR 
library(LDlinkR)      # LD and proxy snps
library(RadialMR)     # Radial MR sensetivity analysis 
library(phenoscanner) # Phenoscaner
library(MVMR)         # Multivarble MR
library(RMVMR)
# library(MendelianRandomization) # Another MVMR 

# Define column types for summary statistics
coltypes = cols(
  ID = col_character(),
  CHROM = col_double(),
  POS = col_double(),
  REF = col_character(),
  ALT = col_character(),
  AF = col_double(),
  TRAIT = col_character(),
  BETA = col_double(),
  SE = col_double(),
  Z = col_double(),
  P = col_double(),
  N = col_double(),
  OR = col_double(),
  OR_L95 = col_double(),
  OR_U95 = col_double(),
  DIR = col_character(),
  G1000_ID = col_character(),
  G1000_VARIANT = col_character(),
  DBSNP_ID = col_character(),
  DBSNP_VARIANT = col_character(),
  OLD_ID = col_character(),
  OLD_VARIANT = col_character()
)

```

```{r get_proxies}
## Function for extracting proxy snps using LDLink
munge_proxies <- function(LDLink_file, outcome, outcome_clump){
  LDLink_file_path <- LDLink_file
  proxy_snps <- read_tsv(LDLink_file_path, skip = 1, col_names = F) %>%
    rename(id = X1, func = X2, proxy_snp = X3, coord = X4, alleles = X5, maf = X6, 
           distance = X7, dprime = X8, rsq = X9, correlated_alleles = X10, FORGEdb = X11, RegulomeDB = X12) %>%
    separate(coord, c('chr', 'pos'), sep = ":") %>%
    mutate(snp = ifelse(id == 1, proxy_snp, NA), 
           chr = str_replace(chr, 'chr', ""), 
           chr = as.numeric(chr), 
           pos = as.numeric(pos)) %>%
    fill(snp, .direction = 'down') %>%
    relocate(snp, .before = proxy_snp) %>%
    dplyr::select(-id, -func, -FORGEdb, -RegulomeDB) %>%
    filter(rsq >= 0.8)

# Munge proxy snp and outcome data
proxy_outcome <- left_join(
    proxy_snps, outcome, by = c("proxy_snp" = "SNP")
  ) %>%
    separate(correlated_alleles, c("target_a1.outcome", "proxy_a1.outcome", 
                                   "target_a2.outcome", "proxy_a2.outcome"), sep = ",|=") %>%
    filter(!is.na(chr.outcome)) %>%
    arrange(snp, -rsq, abs(distance)) %>%
    group_by(snp) %>%
    slice(1) %>%
    ungroup() %>%
    mutate(
       proxy.outcome = TRUE,
       target_snp.outcome = snp,
       proxy_snp.outcome = proxy_snp, 
    ) %>% 
    mutate(
         new_effect_allele.outcome = case_when(
          proxy_a1.outcome == effect_allele.outcome & proxy_a2.outcome == other_allele.outcome ~ target_a1.outcome,
          proxy_a2.outcome == effect_allele.outcome & proxy_a1.outcome == other_allele.outcome ~ target_a2.outcome,
          TRUE ~ NA_character_
       ), 
        new_other_allele.outcome = case_when(
          proxy_a1.outcome == effect_allele.outcome & proxy_a2.outcome == other_allele.outcome ~ target_a2.outcome,
          proxy_a2.outcome == effect_allele.outcome & proxy_a1.outcome == other_allele.outcome ~ target_a1.outcome,
          TRUE ~ NA_character_
       ), 
       effect_allele.outcome = new_effect_allele.outcome, 
       other_allele.outcome = new_other_allele.outcome
    ) %>%
    dplyr::select(-proxy_snp, -chr, -pos, -alleles, -maf, -distance, -rsq, -dprime,  
           -new_effect_allele.outcome, -new_other_allele.outcome) %>%
    relocate(target_a1.outcome, proxy_a1.outcome, target_a2.outcome, proxy_a2.outcome, .after = proxy_snp.outcome) %>%
    rename(SNP = snp) %>%
    relocate(SNP, .after = samplesize.outcome)

# Merge outcome and proxy outcomes
outcome_dat <- bind_rows(
  outcome_clump, proxy_outcome
) %>% 
  arrange(chr.outcome, pos.outcome)

outcome_dat
}

```

## Exposure dataset
The exposure GWAS-SS have previously been standardized to common format, however, they need to be converted to the format required by TwoSampleMR. 

### LDL Cholesterol
```{r exposure1}
exposure1_path = "resources/Willer2013ldl.chrall.CPRA_b37.tsv.gz"
exposure1_ss <- read_tsv(exposure1_path, comment = "##", col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N, TRAIT))

# Format data to TwoSampleMR format
exposure1 <- exposure1_ss %>%
  format_data(.,
    type = "exposure",
    snps = NULL,
    header = TRUE,
    phenotype_col = "TRAIT",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    samplesize_col = "N",
    z_col = "Z",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

# Perform LD clumping on SNP data, filter SNPs to make it run faster
exposure1_clump <- exposure1 %>% 
  filter(pval.exposure < 0.01) %>%
  clump_data(.,
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p1 = 1,
  clump_p2 = 1,
  pop = "EUR"
)

exposure1_dat <- filter(exposure1_clump, pval.exposure < 5e-8)
```

### Exposure 2: HDL Cholesterol 

```{r exposure2}
exposure2_path = "resources/Willer2013hdl.chrall.CPRA_b37.tsv.gz"
exposure2_ss <- read_tsv(exposure2_path, comment = "##", col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N, TRAIT))

# Format data to TwoSampleMR format
exposure2 <- exposure2_ss %>%
  format_data(.,
    type = "exposure",
    snps = NULL,
    header = TRUE,
    phenotype_col = "TRAIT",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    samplesize_col = "N",
    z_col = "Z",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

# Perform LD clumping on SNP data, filter SNPs to make it run faster
exposure2_clump <- exposure2 %>% 
  filter(pval.exposure < 0.01) %>%
  clump_data(.,
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p1 = 1,
  clump_p2 = 1,
  pop = "EUR"
)

exposure2_dat <- filter(exposure2_clump, pval.exposure < 5e-8)
```


## Outcome dataset: Alzheiemr's disease
Similary, the outcome dataset needs to be converted to the TwoSampleMR format and the exposures SNPs or their LD proxies will be extracted.
```{r outcome}
outcome_path = "resources/Kunkle2019load_stage123.chrall.CPRA_b37.tsv.gz"
outcome_ss <- read_tsv(outcome_path, comment = "##",  col_types = coltypes, 
                       col_select = c(DBSNP_ID, CHROM, POS, REF, ALT, AF, BETA, SE, Z, P, N, TRAIT))

# Format outcome
outcome <- outcome_ss %>%
  format_data(.,
    type = "outcome",
    snps = NULL,
    header = TRUE,
    phenotype_col = "TRAIT",
    snp_col = "DBSNP_ID",
    beta_col = "BETA",
    se_col = "SE",
    eaf_col = "AF",
    effect_allele_col = "ALT",
    other_allele_col = "REF",
    pval_col = "P",
    samplesize_col = "N",
    z_col = "Z",
    chr_col = "CHROM",
    pos_col = "POS",
    log_pval = FALSE
) %>%
  as_tibble()

```


## Harmonization

### Combine genetic instruments from exposures 
```{r}

mvmr_snps <- bind_rows(
  select(exposure1_dat, SNP), 
  select(exposure2_dat, SNP), 
) %>%
  distinct(SNP) %>%
  pull(SNP)

comb_exp <- bind_rows(
  exposure1 %>% filter(SNP %in% mvmr_snps),
  exposure2 %>% filter(SNP %in% mvmr_snps),
)

comb_exp_out <- bind_rows(
  exposure1 %>% filter(SNP %in% mvmr_snps) %>% 
    rename(pval = pval.exposure, trait = exposure, chr = chr.exposure, pos = pos.exposure), 
  exposure2 %>% filter(SNP %in% mvmr_snps) %>% 
    rename(pval = pval.exposure, trait = exposure, chr = chr.exposure, pos = pos.exposure), 
  outcome %>% filter(SNP %in% mvmr_snps) %>% 
    rename(pval = pval.outcome, trait = outcome, chr = chr.outcome, pos = pos.outcome)
)

## How many SNPs are GWS across the exposures 
comb_exp_wide <- comb_exp_out %>%
   arrange(chr, pos) %>%
   select(SNP, trait, pval) %>%
   pivot_wider(names_from = trait, values_from = pval) %>%
   mutate(pval.exposure = pmin(LDL, HDL, na.rm = T))

comb_exp_wide %>% count(LDL < 5e-8, HDL < 5e-8)
comb_exp_wide %>% count(LDL < 5e-8, HDL < 5e-8, AD < 5e-8)

## Do another round of LD clumping 
comb_exp_clump <- comb_exp %>% 
  clump_data(.,
  clump_kb = 10000,
  clump_r2 = 0.001,
  clump_p1 = 1,
  clump_p2 = 1,
  pop = "EUR"
)

comb_exp_clump %>% count(LDL < 5e-8, HDL < 5e-8)


comb_exp <- comb_exp %>% filter(SNP %in% comb_exp_clump$SNP) 

```

### Extract Outcomes
```{r}

# extract exposure SNPs present in outcome
outcome_clump <- semi_join(
  outcome, comb_exp, by = "SNP"
)

# Exposure SNPs not present in outomce
outcome_wo_snps <- anti_join(
  comb_exp, outcome, by = "SNP"
) %>%
  distinct(SNP)

# Use LDLinkR to identify proxy snps
LDproxy_batch(outcome_wo_snps, 
        pop = "CEU", 
        r2d = "r2", 
        token = 'a6deee62cc4a', 
        append = TRUE,
        genome_build = "grch37")
system("mv combined_query_snp_list_grch37.txt data/mvmr_proxy_snps.txt")

# Munge proxy snp file
outcome_dat <- munge_proxies("data/mvmr_proxy_snps.txt", outcome, outcome_clump)

```

```{r VennDiagram}
library(nVennR)
## GWS SNPs
set1 <- list(pull(exposure1_dat, SNP))
set2 <- list(pull(exposure2_dat, SNP))
myNV <- plotVenn(list(set1, set2), sNames=c("LDL", "HDL"), 
                 showPlot = F)
showSVG(myNV, opacity=0.2, outFile = "results/plots/mvmr_gws_venn.svg")

## All SNPs 
set1 <- list(exposure1 %>% filter(SNP %in% mvmr_snps, pval.exposure < 5e-8) %>% pull(SNP))
set2 <- list(exposure2 %>% filter(SNP %in% mvmr_snps, pval.exposure < 5e-8) %>% pull(SNP))
set3 <- list(outcome %>% filter(SNP %in% mvmr_snps) %>% pull(SNP))
myNV <- plotVenn(list(set1, set2, set3), sNames=c("LDL", "HDL", "AD"), 
                 showPlot = F)
showSVG(myNV, opacity=0.2, outFile = "results/plots/mvmr_venn.svg")


```

### Harmonize
```{r}
# Univariate MR 
mrdat <- harmonise_data(comb_exp, outcome_dat) %>%
  as_tibble() %>%
  filter(pval.exposure < 5e-8)

mvdat <- mv_harmonise_data(comb_exp, outcome_dat)
```

```{r export_harmonized_dat}
write_csv(mrdat, 'data/harmonized_mvmr_uni.csv')
write_rds(mvdat, 'data/harmonized_mvmr.rds')

```

## Results 

### TwoSampleMR
```{r TwoSampleMR_MR}
## Univariate MR
mr_res <- mr(mrdat, method_list = c("mr_ivw_fe", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode"))

# Heterogeneity statistics 
het_res <- mr_heterogeneity(mrdat, method_list = c("mr_egger_regression", "mr_ivw"))

# Horizontal pleitropy
plei_res <- mr_pleiotropy_test(mrdat)

# Leave-one-out analysis 
res_loo <- mr_leaveoneout(mrdat, method = mr_ivw_fe) %>% as_tibble()

# Single SNP anlaysis 
res_single <- mr_singlesnp(mrdat, all_method = c("mr_ivw_fe", "mr_egger_regression", "mr_weighted_median", "mr_weighted_mode")) %>% as_tibble()

# Radial MR 
radial_dat <- mrdat %>% filter(mr_keep == T) %>% dat_to_RadialMR()
radial_res <- map(radial_dat, function(x){
    ivw_radial(x, alpha = 0.05/nrow(x))
  }
)


```

```{r plots}
scatter_p <- mr_scatter_plot(mr_res, mrdat)

funnel_p <- mr_funnel_plot(res_single)

radial_p <-  map(radial_res, function(x){
     plot_radial(x, radial_scale = F, show_outliers = T)
  }
)

radial_p <- plot_radial(radial_res$LDL.AD, radial_scale = F, show_outliers = T)
radial_p

joint_mr_p <- cowplot::plot_grid(
  scatter_p[[1]] + labs(title = "HDL") + theme_bw() + theme(legend.position = 'none', text = element_text(size = 8)),
  funnel_p[[1]] + labs(title = "HDL") + theme_bw() + theme(legend.position = 'none', text = element_text(size = 8)), 
  radial_p[[1]] + labs(title = "HDL") + coord_fixed(ratio=0.25) + theme(legend.position = 'none', text = element_text(size = 6)), 
  scatter_p[[2]] + labs(title = "LDL") + theme_bw() + theme(legend.position = 'none', text = element_text(size = 8)), 
  funnel_p[[2]] + labs(title = "LDL") + theme_bw() + theme(legend.position = 'none', text = element_text(size = 8)),
  radial_p[[2]] + labs(title = "LDL") + coord_fixed(ratio=0.25) + theme(legend.position = 'none', text = element_text(size = 6))
  )


legend_scatter <- cowplot::get_legend(
  # create some space to the left of the legend
  scatter_p[[1]] + theme(legend.box.margin = margin(12, 12, 12, 12))
)

# add the legend to the row we made earlier. Give it one-third of 
# the width of one plot (via rel_widths).
scatter_out.p <- cowplot::plot_grid(joint_mr_p, legend_scatter, rel_heights = c(4, 1), nrow = 2)

radial_p[[2]] + labs(title = "LDL") + theme(legend.position = 'none', text = element_text(size = 6)) + coord_fixed(ratio=0.5)
```


```{r TwoSampleMR_MVMR}
## Multivariable MR
mvmr_res <- mv_multiple(mvdat, plots = T)
mvmr_res$result

```

```{r export_twosamplemr}
write_csv(mr_res, 'results/mvmr_uni_res.csv')
write_csv(het_res, 'results/mvmr_heterogenity_res.csv')
write_csv(plei_res, 'results/mvmr_pleiotropy_res.csv')
write_csv(mvmr_res$result, 'results/mvmr_res.csv')

ggsave('results/plots/mr_scatter_funnel.png', plot = scatter_out.p, units = 'in', height = 4, width = 9)


## GT 
mr_res %>%
  select(-id.exposure, -id.outcome) %>%
  gt() %>%
  fmt_number(
    columns = c("b", "se")
  ) %>%
  fmt_number(
    columns = pval,
    rows = pval > 0.001,
    decimals = 3
  ) %>% 
  fmt_scientific(
    columns = pval,
    rows = pval <= 0.001,
    decimals = 1
  ) %>%
  gtsave(., 'results/tables/mvmr_uni_res.png')

### Heterogenity
het_res %>%
  select(-id.exposure, -id.outcome) %>%
  gt() %>%
  fmt_number(
    columns = Q
  ) %>%
  fmt_number(
    columns = Q_pval,
    rows = Q_pval > 0.001,
    decimals = 3
  ) %>% 
  fmt_scientific(
    columns = Q_pval,
    rows = Q_pval <= 0.001,
    decimals = 1
  ) %>%
  gtsave(., 'results/tables/mvmr_uni_het.png')  

### Pleiotropy
plei_res %>%
  select(-id.exposure, -id.outcome) %>%
  gt() %>%
  fmt_number(
    columns = c('egger_intercept', 'se')
  ) %>%
  fmt_number(
    columns = pval,
    rows = pval > 0.001,
    decimals = 3
  ) %>% 
  fmt_scientific(
    columns = pval,
    rows = pval <= 0.001,
    decimals = 1
  ) %>%
  gtsave(., 'results/tables/mvmr_uni_plt.png')  
```

### MVMR package

```{r MVMR}
mvmr_dat <- cbind(mvdat$exposure_beta, 
        mvdat$exposure_se, 
        mvdat$outcome_beta, 
        mvdat$outcome_se
      ) %>%
      as.data.frame() %>%
      rownames_to_column() %>%
      as.matrix()

F.data <- format_mvmr(BXGs = mvmr_dat[,c(2,3)],
                    BYG = mvmr_dat[,6],
                    seBXGs = mvmr_dat[,c(4,5)],
                    seBYG = mvmr_dat[,7],
                    RSID = mvmr_dat[,1])

sres <- strength_mvmr(r_input = F.data, gencov = 0)
pres <- pleiotropy_mvmr(r_input = F.data, gencov = 0)
res <- ivw_mvmr(r_input = F.data)

## RMVMR
F.data_rmvmr <- format_rmvmr(BXGs = mvmr_dat[,c(2,3)],
                    BYG = mvmr_dat[,6],
                    seBXGs = mvmr_dat[,c(4,5)],
                    seBYG = mvmr_dat[,7],
                    RSID = mvmr_dat[,1])


res_rmvrm <- ivw_rmvmr(F.data_rmvmr)
sres_rmvrm <- strength_rmvmr(F.data_rmvmr, gencov = 0)
rmvrm_p <- plot_rmvmr(F.data_rmvmr, res_rmvrm)

```

```{r export_RMVMR}
mvmr_out <- bind_rows(
  res_rmvrm$coef %>% 
    as.data.frame() %>%
    rownames_to_column() %>%
    as_tibble() %>% 
    janitor::clean_names() %>%
    rename(exposure = rowname, b = estimate, se = std_error, p = pr_t) %>%
    mutate(
      exposure = ifelse(exposure == 'exposure1', 'HDL', 'LDL'), 
      outcome = 'AD', 
      method = 'Radial-MVMR'
    ) %>%
    relocate(outcome, method, .after = exposure), 

  res %>% 
    as.data.frame() %>%
    rownames_to_column() %>%
    as_tibble() %>% 
    janitor::clean_names() %>%
    rename(exposure = rowname, b = estimate, se = std_error, p = pr_t) %>%
    mutate(
      exposure = ifelse(exposure == 'exposure1', 'HDL', 'LDL'), 
      outcome = 'AD',
      method = 'IVW-MVMR'
    ) %>%
    relocate(outcome, method, .after = exposure)

)

write_csv(mvmr_out, 'results/rmvmr_res.csv')

f_stat <- bind_rows(
  sres_rmvrm$f, 
  sres
)

ggsave('results/plots/rmvmr_scatter_funnel.png', plot = rmvrm_p[[1]], units = 'in', height = 4, width = 9)


```

### MendelianRandomization & MVMRmode Package
```{r MendelianRandomization}

MRMVInputObject <- MendelianRandomization::mr_mvinput(
  bx = mvdat$exposure_beta, 
  bxse = mvdat$exposure_se,
  by = mvdat$outcome_beta,
  byse = mvdat$outcome_se,
  snps = rownames(mvdat$exposure_beta), 
  exposure = c('HDL', 'LDL'), 
  outcome = 'AD'
  )

MRMV_ivw <- MendelianRandomization::mr_mvivw(MRMVInputObject, 
                       model="default", 
                       correl=FALSE, 
                       distribution="normal",
                       alpha = 0.05)

MRMV_ivw_df <- cbind(MRMV_ivw@Exposure, MRMV_ivw@Outcome, MRMV_ivw@Estimate, MRMV_ivw@StdError, MRMV_ivw@Pvalue) %>%
  as_tibble() %>%
  magrittr::set_colnames(c("Exposure", "Outcome", "b", "se", "p")) %>%
  mutate(method = "MRMV-IVW")

MRMV_egger <- MendelianRandomization::mr_mvegger(MRMVInputObject, orientate = 1, correl = FALSE, distribution = "normal", alpha = 0.05)

MRMV_egger_df <- cbind(MRMV_egger@Exposure, MRMV_egger@Outcome, MRMV_egger@Estimate, MRMV_egger@StdError.Est, MRMV_egger@Pvalue.Est) %>%
  as_tibble() %>%
  magrittr::set_colnames(c("Exposure", "Outcome", "b", "se", "p")) %>%
  mutate(method = "MRMV-Egger")


MRMV_wme <- MendelianRandomization::mr_mvmedian( MRMVInputObject, distribution = "normal", alpha = 0.05, iterations = 10000, seed = 314159265 )

MRMV_wme_df <- cbind(MRMV_wme@Exposure, MRMV_wme@Outcome, MRMV_wme@Estimate, MRMV_wme@StdError, MRMV_wme@Pvalue) %>%
  as_tibble() %>%
  magrittr::set_colnames(c("Exposure", "Outcome", "b", "se", "p")) %>%
  mutate(method = "MRMV-WME")


MRMV_wbe_res <- MVMRmode::mv_mrmode(Bout = mvdat$outcome_beta, 
                    Bexp = mvdat$exposure_beta, 
                    SEout = mvdat$outcome_se, 
                    SEexp = mvdat$exposure_se, 
                    Psi=0, CIMin = NA, CIMax = NA, CIStep = 0.001, alpha = 0.05, 
                    residual="IVW",Mode="CM",weighting = "weighted", stderror = "simple", 
                    phi = 1,distribution = "normal",iterations = 10000)

MRMV_wbe_df <- as_tibble(MRMV_wbe_res) %>%
  rename(b = Estimate, p = Pvalue) %>%
  mutate(Exposure = c("LDL", "HDL"), 
         Outcome = "AD", 
         method = 'MRMV_WBE') %>%
  mutate_if(is.double, as.character)


```

```{r MRMV_out}


MRMV_out <- bind_rows(
    MRMV_ivw_df, MRMV_egger_df, MRMV_wme_df, MRMV_wbe_df
  ) %>%
  relocate(method, .after = Outcome) %>%
  select(Exposure, Outcome, method, b, se, p) %>%
  mutate_at(vars(b, se, p), as.double)

write_csv(MRMV_out, 'results/mvmr_sensetivity_results.csv')

MRMV_out %>%
  pivot_wider(names_from = Exposure, values_from = c('b', 'se', 'p')) %>%
  relocate(se_HDL, p_HDL, .after = b_HDL) %>%
  gt() %>%
  fmt_number(
    columns = c("b_HDL", "se_HDL", "b_LDL", "se_LDL")
  ) %>% 
  fmt_number(
    columns = c("p_LDL"),
    rows =  p_LDL > 0.001,
    decimals = 3
  ) %>% 
  fmt_scientific(
    columns = c("p_LDL"),
    rows = p_LDL <= 0.001,
    decimals = 1
  ) %>%
  fmt_number(
    columns = c("p_HDL"),
    rows = p_HDL > 0.001 ,
    decimals = 3
  ) %>% 
  fmt_scientific(
    columns = c("p_HDL"),
    rows = p_HDL <= 0.001,
    decimals = 1
  ) %>%
  tab_spanner(
    label = "HDL",
    columns = c("b_HDL", "se_HDL", "p_HDL")
  ) %>%
  tab_spanner(
    label = "LDL",
    columns = c("b_LDL", "se_LDL", "p_LDL")
  ) %>%
  cols_label(
    b_LDL = 'b',
    se_LDL = 'se',
    p_LDL = 'p',
    b_HDL = 'b',
    se_HDL = 'se',
    p_HDL = 'p',
  ) %>%
  gtsave(., 'results/tables/mvmr_res.png')

```





























