# Standardizing and harmonizing GWAS SumStats

Two-sample MR methods require the use of summary statistics from genome-wide association studies, including single nucleotide polymorphisms (SNPs), beta coefficients, standard errors, p-values, and allele frequencies. However, the historical lack of standards for data content and file formats in GWAS summary statistics has resulted in heterogeneous data sets. To address this issue, standardizing and harmonizing the GWAS summary statistics is crucial before conducting MR analyses. The [GWAS Catalog](https://www.ebi.ac.uk/gwas/home) and [OpenGWAS](https://gwas.mrcieu.ac.uk/) platforms have developed formats such as GWAS-SSF (Hayhurst et al. 2022) and GWAS-VCF (Lyon et al. 2021) to facilitate sharing of GWAS SumStats. Tools like [MungeSumstats](https://neurogenomics.github.io/MungeSumstats/index.html) (Murphy et al 2021) and GWAS2VCF (Lyon et al. 2021) are available that provide rapid standardization and quality control of GWAS SumStats.
